File: backend/app/services/progress_service.py
from sqlmodel import Session, select, and_
from ..database import engine
from ..models import UserProgress, User
from typing import Set, Optional
from datetime import datetime, timedelta

class ProgressService:
    def mark_completed(self, username: str, question_id: str, code: str = None):
        username = username.strip()
        question_id = question_id.strip()
        with Session(engine) as session:
            # Check if exists (handle duplicates robustly)
            statement = select(UserProgress).where(
                and_(UserProgress.username == username, UserProgress.question_id == question_id)
            )
            results = session.exec(statement).all()
            
            is_new_completion = False
            if not results:
                # No existing progress for this user/question
                progress = UserProgress(username=username, question_id=question_id, status="completed", user_code=code)
                session.add(progress)
                is_new_completion = True
            else:
                # Check if it was already completed to avoid double counting streak
                was_already_completed = any(p.status == "completed" for p in results)
                if not was_already_completed:
                    is_new_completion = True

                # Update first one to completed, delete others if any (to fix existing duplicates)
                results[0].status = "completed"
                if code: results[0].user_code = code
                results[0].updated_at = datetime.utcnow()
                session.add(results[0])
                
                for p in results[1:]:
                    session.delete(p)
            
            if is_new_completion:
                self._update_streak(session, username)
            session.commit()

    def mark_attempted(self, username: str, question_id: str, code: str = None):
        username = username.strip()
        question_id = question_id.strip()
        with Session(engine) as session:
            statement = select(UserProgress).where(
                and_(UserProgress.username == username, UserProgress.question_id == question_id)
            )
            results = session.exec(statement).all()
            
            if not results:
                progress = UserProgress(username=username, question_id=question_id, status="attempted", user_code=code)
                session.add(progress)
            else:
                # If ANY is completed, preserve it. Just update code if needed.
                is_completed = any(p.status == "completed" for p in results)
                # Use the first one as primary
                primary = results[0]
                if not is_completed:
                    primary.status = "attempted"
                
                # Update code, allowing None (failed attempt) to clear it
                primary.user_code = code
                
                primary.updated_at = datetime.utcnow()
                session.add(primary)
                
                # Cleanup duplicates
                for p in results[1:]:
                    session.delete(p)
            
            session.commit()

    def get_progress(self, username: str) -> dict:
        username = username.strip()
        with Session(engine) as session:
            statement = select(UserProgress).where(UserProgress.username == username)
            results = session.exec(statement).all()
            
            # Use a dict to ensure only one status per question_id
            # Preference: completed > attempted
            status_map = {}
            for p in results:
                qid = p.question_id.strip()
                if qid not in status_map or p.status == "completed":
                    status_map[qid] = p.status
            
            return {
                "completed": {qid for qid, status in status_map.items() if status == "completed"},
                "attempted": {qid for qid, status in status_map.items() if status == "attempted"}
            }

    def get_user_code(self, username: str, question_id: str) -> Optional[str]:
        username = username.strip()
        question_id = question_id.strip()
        with Session(engine) as session:
            statement = select(UserProgress).where(
                and_(UserProgress.username == username, UserProgress.question_id == question_id)
            ).order_by(UserProgress.updated_at.desc())
            progress = session.exec(statement).first()
            return progress.user_code if progress else None

    def _update_streak(self, session: Session, username: str):
        statement = select(User).where(User.username == username)
        user = session.exec(statement).first()
        if not user:
            return

        # Simplified "Solve Streak" logic: Always increment on successful completion
        # This allows users to gain streak points for every question they answer,
        # which is more rewarding for a practice session than a strict "daily" counter.
        now = datetime.utcnow()
        user.current_streak += 1
        user.last_completed_at = now
        session.add(user)

    def refresh_streak(self, username: str):
        """Check if streak should be reset due to inactivity."""
        with Session(engine) as session:
            statement = select(User).where(User.username == username)
            user = session.exec(statement).first()
            if not user or not user.last_completed_at:
                return

            now = datetime.utcnow()
            today = now.date()
            last_date = user.last_completed_at.date()

            if last_date < today - timedelta(days=1):
                # Skipped yesterday, reset streak to 0
                user.current_streak = 0
                session.add(user)
                session.commit()

progress_service = ProgressService()

File: backend/src/progress/progress.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service.js';

@Injectable()
export class ProgressService {
  constructor(private readonly prisma: PrismaService) {}

  async markCompleted(
    userId: number,
    questionId: string,
    code?: string,
  ): Promise<void> {
    questionId = questionId.trim();

    await this.prisma.userProgress.upsert({
      where: { user_id_question_id: { user_id: userId, question_id: questionId } },
      update: {
        status: 'completed',
        ...(code ? { user_code: code } : {}),
        updated_at: new Date(),
      },
      create: {
        user_id: userId,
        question_id: questionId,
        status: 'completed',
        user_code: code || null,
      },
    });
  }

  async markAttempted(
    userId: number,
    questionId: string,
    code?: string,
  ): Promise<void> {
    questionId = questionId.trim();

    const existing = await this.prisma.userProgress.findUnique({
      where: { user_id_question_id: { user_id: userId, question_id: questionId } },
    });

    if (!existing) {
      await this.prisma.userProgress.create({
        data: {
          user_id: userId,
          question_id: questionId,
          status: 'attempted',
          user_code: code || null,
        },
      });
    } else {
      // Preserve completed status
      await this.prisma.userProgress.update({
        where: { id: existing.id },
        data: {
          ...(existing.status !== 'completed' ? { status: 'attempted' } : {}),
          user_code: code ?? existing.user_code,
          updated_at: new Date(),
        },
      });
    }
  }

  async getProgress(
    userId: number,
  ): Promise<{ completed: Set<string>; attempted: Set<string> }> {
    const records = await this.prisma.userProgress.findMany({
      where: { user_id: userId },
    });

    const completed = new Set<string>();
    const attempted = new Set<string>();

    for (const p of records) {
      const qid = p.question_id.trim();
      if (p.status === 'completed') {
        completed.add(qid);
      } else if (p.status === 'attempted') {
        attempted.add(qid);
      }
    }

    return { completed, attempted };
  }

  async getUserCode(
    userId: number,
    questionId: string,
  ): Promise<string | null> {
    questionId = questionId.trim();

    const progress = await this.prisma.userProgress.findUnique({
      where: { user_id_question_id: { user_id: userId, question_id: questionId } },
    });

    return progress?.user_code ?? null;
  }
}
